File: 2011/CCC 2012 S3/S3.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<string>> base = {
    {"empty", "crystal", "crystal", "crystal", "empty"},
    {"empty", "continue", "crystal", "continue", "empty"},
    {"empty", "empty", "continue", "empty", "empty"},
    {"empty", "empty", "empty", "empty", "empty"},
    {"empty", "empty", "empty", "empty", "empty"}
};

vector<vector<string>> base2 = {
    {"empty", "crystal", "crystal", "crystal", "empty"},
    {"empty", "empty", "crystal", "empty", "empty"},
    {"empty", "empty", "empty", "empty", "empty"},
    {"empty", "empty", "empty", "empty", "empty"},
    {"empty", "empty", "empty", "empty", "empty"}
};

string recur(int x, int y, int cur_mag){
    if (cur_mag == 1){
        return base2[y][x];
    } else{
        int localX = floor(((x)/(double)pow (5, cur_mag - 1))); // 0 - 4
        int localY = floor(((y)/(double)pow (5, cur_mag - 1))); // 0 - 4
        
        if (base[localY][localX] == "crystal"){
            return "crystal";
        } else if (base[localY][localX] == "empty"){
            return "empty";
        } else{
            int newX = x - (localX) * pow(5, cur_mag-1);
            int newY = y - (localY) * pow(5, cur_mag-1);
            return recur(newX, newY, cur_mag - 1);
        }
    }
}

void solve(){
    int m, x ,y;
    cin >> m >> x >> y;
    cout << recur(x,y,m) << "\n";
}

int main(){
    int n;

    cin >> n;

    for (int i = 0; i < n; i++){
        solve();
    }
}

File: 2011/CCC 2012 S3/S3.py
import math

t = int(input())

base1 = [
    ["empty", "crystal", "crystal", "crystal", "empty",],
    ["empty", "empty", "crystal", "empty", "empty",],
    ["empty", "empty", "empty", "empty", "empty",],
    ["empty", "empty", "empty", "empty", "empty",],
    ["empty", "empty", "empty", "empty", "empty",]
]

base2 = [
    ["empty", "crystal", "crystal", "crystal", "empty",],
    ["empty", "continue", "crystal", "continue", "empty",],
    ["empty", "empty", "continue", "empty", "empty",],
    ["empty", "empty", "empty", "empty", "empty",],
    ["empty", "empty", "empty", "empty", "empty",]
]

def recur(x, y, level):
    if (level == 1):
        return base1[y][x]
    else:
        localX = math.floor(x/(5 ** (level - 1)))
        localY = math.floor(y/(5 ** (level - 1)))
        if (base2[localY][localX] == "continue"):
            newX = x - (localX) * (5 ** (level-1))
            newY = y - (localY) * (5 ** (level-1))
            return recur(newX, newY, level-1)
        else:
            return base1[localY][localX]
        



def solve():
    array = [int(x) for x in input().split()]

    level = array[0]
    x = array[1]
    y = array[2]

    print(recur(x ,y, level))

    return 0

for i in range(t):
    solve()

    